### PR TITLE
[JS] feat: TeamsBotSsoPrompt based SSO support

### PR DESCRIPTION
**Pull request by blackchoey**
## Linked issues

closes: #781 #782

## Details

1. Refactored current authentication design to accept different implementation.
    * Extracted common logic to a base class
    * Messaging extension authentication method no longer accepts a settings parameter, it's already provided in constructor
    * Split `runDialog` to `runDialog` and `continueDialog`, to better handle the events
    * Minor update to logic around Messaging extension magic code to make the flow more clear
2. Add TeamsBotSsoPrompt based SSO implementation.
3. Add test cases

TODO:
1. Add samples (will create another PR)

## Attestation Checklist

- [ ] My code follows the style guidelines of this project

- I have checked for/fixed spelling, linting, and other errors
- I have commented my code for clarity
- I have made corresponding changes to the documentation (we use [TypeDoc](https://typedoc.org/) to document our code)
- My changes generate no new warnings
- I have added tests that validates my changes, and provides sufficient test coverage. I have tested with:
  - Local testing
  - E2E testing in Teams
- New and existing unit tests pass locally with my changes

### CODE REVIEW

File: js/packages/teams-ai/src/authentication/Authentication.ts
Comment (corinagum) [inline js/packages/teams-ai/src/authentication/Authentication.ts:168]: Here, the return type can be promise of undefined, but on line 178, it can be null. Is there a reason for the difference?

File: js/packages/teams-ai/src/authentication/MessagingExtensionAuthenticationBase.ts
Comment (corinagum) [inline js/packages/teams-ai/src/authentication/MessagingExtensionAuthenticationBase.ts:7]: Isn't the naming `MessageExtension`? Or is it different for authentication?

Comment (corinagum) [reply js/packages/teams-ai/src/authentication/MessagingExtensionAuthenticationBase.ts:7]: [Documentation](https://learn.microsoft.com/en-us/microsoftteams/platform/messaging-extensions/how-to/add-authentication) would indicate the correct term is `MessageExtension`:
image

Could we rename away from `MessagingExtension`?

Comment (corinagum) [reply js/packages/teams-ai/src/authentication/MessagingExtensionAuthenticationBase.ts:7]: Uhhh nevermind. Botbuilder has imports that are also `MessagingExtension` instead of `MessageExtension` so there's no way to fully fix the inconsistency.

File: js/packages/teams-ai/src/authentication/OAuthPromptBotAuthentication.ts
Comment (corinagum) [inline js/packages/teams-ai/src/authentication/OAuthPromptBotAuthentication.ts:16]:
```suggestion
        oAuthPromptSettings: OAuthPromptSettings, // Child classes will have different types for this
```
?

File: js/packages/teams-ai/src/authentication/Authentication.ts
Comment (blackchoey) [reply js/packages/teams-ai/src/authentication/Authentication.ts:168]: In my mind, a function should return null because it clearly tells something does not have a value. And some library (like MSAL) also follows this rule. So I will try to return null if possible, like what I'm doing on line 178.
But I met an external function that declares loginHint parameter as optional (xxxFunction(loginHint?: string), null is not accepted in this case. Since `getLoginHint` is just a private function, I returned undefined to reduce the coding complexity.

I think this is a coding style thing. If there's some style I need to follow, I will update the PR to satisfy the style.

File: js/packages/teams-ai/src/authentication/MessagingExtensionAuthenticationBase.ts
Comment (corinagum) [inline js/packages/teams-ai/src/authentication/MessagingExtensionAuthenticationBase.ts:8]: Sorry for the hassle, but please rename this file and **new** methods/variables/etc with MessageExtension instead of MessagingExtension. We will try to stay consistent about this from now on.

File: js/packages/teams-ai/src/authentication/OAuthPromptMessagingExtensionAuthentication.ts
Comment (corinagum) [inline js/packages/teams-ai/src/authentication/OAuthPromptMessagingExtensionAuthentication.ts:9]: Please rename to `OAuthPromptMessageExtensionAuthentication`

File: js/packages/teams-ai/src/authentication/Authentication.ts
Comment (corinagum) [reply js/packages/teams-ai/src/authentication/Authentication.ts:168]: thanks for the explanation - this looks good as-is to me :)

File: js/packages/teams-ai/src/authentication/TeamsSsoMessagingExtensionAuthentication.ts
Comment (corinagum) [inline js/packages/teams-ai/src/authentication/TeamsSsoMessagingExtensionAuthentication.ts:12]: Please rename to `TeamsSsoMessageExtensionAuthentication` (same with any non-external package instances that say MessagingExtension`

**Review by corinagum — DISMISSED** [review]
LGTM! Sorry for the hassle but if you could please do the rename to `...MessageExtension...`, this is the correct terminology and we need to stay as consistent about it as possible, even though some instances must stay as-is.

If you have the permissions to merge after the change, feel free to do so. Otherwise, feel free to ping me any time and I will merge in asap.

**Review by singhk97 — APPROVED** [review]

File: js/packages/teams-ai/src/TurnState.ts
Comment (blackchoey) [reply js/packages/teams-ai/src/TurnState.ts:64]: This is used by our library internally. Our library will save the state after processing every request. This will cause ETag conflict when processing the duplicate token exchange activities. So I introduced this flag to let our library be able to skip state saving for duplicated token exchange activity.

I think it can also be adopted by oAuthPrompt. But since we're using the token exchange middleware from botbuilder to handle the duplicate for oAuthPrompt, it's currently only used for TeamsSsoPrompt. I choose the current implementation to reduce the token exchange requests.

**Review by corinagum — DISMISSED** [review]

File: js/packages/teams-ai/src/TurnState.ts
Comment (singhk97) [reply js/packages/teams-ai/src/TurnState.ts:69]: Since it's used internally in the library we should ideally obscure this from the end user, just like how state the Bot Authentication state is stored. Does this method scale to multiple concurrent login requests from different users?

**Review by singhk97 — DISMISSED** [review]

**Comment by singhk97** [comment]
I think that `OAuthPrompt` prefix to `OAuthPromptMessageExtensionAuthentication` makes the class name to long. I think that "Prompt" in the name is superfluous because it does inform much other than we're using "OAuthPrompt" dialog under the hood. 

I propose we change "OAuthPrompt" to just "OAuth". And TeamsSSOPrompt to "TeamsSSO" to keep things simple.

**Comment by singhk97** [comment]
Also, let's pick a consistent variable/property naming for "OAuth". Let's stick with `oauth` or `oAuth` but not both. @blackchoey it's your call

**Review by lilyydu — APPROVED** [review]

**Pull request merged**
